### ISSUE TITLE
Confusing error position with new stricter type check and object spread operator

### ISSUE THREAD

**Issue by CnSimonChan** [Microsoft/TypeScript#16473]
**TypeScript Version:**  2.4.0

**Code**

```` ts

interface A {
    a: string;
    b: string;
}

// A
const a0: A = {
    a: "a",
    b: "b",
    c: "c", // (1)
};

// B
const b1 = {
    a: "a",
    b: "b",
    c: "c", // (4)
};

const a1: A = { ...b1 /* (2) */ };

// C
interface B {
    a: string;
    b: string;
    c: string; // (5)
}

const b2: B = {
    a: "a",
    b: "b",
    c: "c",
};

const a2: A = { ...b2 /* (3) */ };

````

**Expected behavior:**

Type incompatible error at `c: "c", // (1)`, `...b1 /* (2) */` and `...b2 /* (3) */`

````
Type '{ a: string; b: string; c: string; }' is not assignable to type 'A'.
  Object literal may only specify known properties, and 'c' does not exist in type 'A'.
````

**Actual behavior:**

Error at  `c: "c", // (1)`, `c: "c", // (4)` and `c: string; // (5)`. The later two are confusing.

----

I really don't know how can this check improve code quality. I can always assign the spread result to a temp local variable and then assign it to the final value, TypeScript doesn't care about this at all:

```` ts

const b1 = {
    a: "a",
    b: "b",
    c: "c",
};

const a1: A = b1;

````

**Comment by sandersn** [comment]
Fix is up at #16530

**Comment by CnSimonChan** [comment]
@sandersn Thanks for the fix, but I prefer the error to be on the spread operator, so if I have a long object literal or with multiple spread operators, I can easily know which one caused the error. I think it's more reasonable.

**Comment by sandersn** [comment]
Yes, the ideal fix reports errors on the spread itself. However, it would require some far-reaching changes to the compiler to make this happen, so I decided to make the 2-line change for 2.4.

Here's why it's so complex to report errors on the spreads themselves: spread operations happen during `checkObjectLiteral`, when the anonymous type is first created. Spread properties are dumped into the anonymous object type and look like any other properties. `hasExcessProperties` operates on types, but happens during assignability checking, long after the types have been created.

But surely the properties of a type remember that they came from a spread, right? Well, actually they don't. For efficiency, spread properties are put intact into the containing anonymous object type. (Intersections and unions work almost the same way.) New properties are only created if something needs to change, like removing `readonly`. That's why the errors in this bug were so horribly confusing &mdash; they pointed all the way back to the original declaration of the properties.

So, in order to find the containing spread, you'd (1) need to create fresh, synthetic symbols for *every* spread property, not just readonly ones, or (2) re-do a lot of the spread work to figure out which symbols came from which spread. Spreads could be arbitrarily tricky, so you really would need to re-run the original logic in error-detection mode. Consider spreading duplicate properties:

```ts
let o = { extra: 1 }
let spread: { a: number } = { extra: 2, ...o, ...o }
```

The error span should be under the last `...o` here, but to know that you need to imitate (or re-use) the right-most scanning style of `createSpreadType+checkObjectLiteral`.

(2) is probably the right fix because it's only slow in the error case. But it's a lot of work for a rare case. At least, we believe (with no data for backup!) that long object literals with multiple spreads are rare compared to simple two-element literals with one or two spreads. The best way to get this changed is to get upvotes for the complete fix or show data that the simple fix in 2.4 isn't sufficient.
